### PR TITLE
Fix #43906: Completing support for 256th, 512th and 1024th durations

### PR DESCRIPTION
**Pull request by Jojo-Schmitz**
solves https://musescore.org/en/node/43906

I've now also had a look at all the changes the older PR #2677 made and applied them to this PR too (with needed adjustments of course).

Any ideas what the shortest durations are that Capella and Overture support? As far as I can tell Finale supports up to 128th and Sibelius up to 512th, not sure about Dorico, but we don't import any of those, so that'd not be too relevant,  Capella and Overture though are, Capella and Overture  import currently work up to 256th. I've been told though that Capella's shortest duration is 128th?

### CODE REVIEW

File: libmscore/stafftype.cpp [hunk 1]
```diff
@@ -1137,6 +1141,10 @@ bool TablatureDurationFont::read(XmlReader& e)
                         displayValue[int(TabVal::VAL_128)] = chr;
                   else if (val == "256")
                         displayValue[int(TabVal::VAL_256)] = chr;
+//                  else if (val == "512")
+//                        displayValue[int(TabVal::VAL_512)] = chr;
+//                  else if (val == "1024")
+//                        displayValue[int(TabVal::VAL_1024)] = chr;
```
Comment (Howard-C) [inline libmscore/stafftype.cpp]: Why commented out?

Comment (Jojo-Schmitz) [reply libmscore/stafftype.cpp]: Because it is still WIP and `TabVal::VAL_512` and `TabVal::VAL_1024` don't exist (yet) ;-)

Comment (Jojo-Schmitz) [reply libmscore/stafftype.cpp]: There you go ;-)

File: mscore/icons.h [hunk 2]
```diff
@@ -24,9 +24,9 @@ namespace Ms {
 
 extern void genIcons();
 
-enum class Icons : signed char { Invalid_ICON = -1,
+enum class Icons : short { Invalid_ICON = -1,
```
Comment (Jojo-Schmitz) [inline mscore/icons.h]: -1 to 130 no longer fits `signed char`

File: libmscore/tempotext.cpp [hunk 3]
```diff
@@ -120,6 +120,9 @@ static const TempoPattern tp[] = {
       TempoPattern("\uECA0",                     1.0/7.5,    TDuration::DurationType::V_BREVE),      // double whole
       TempoPattern("\uECAD",                     1.0/960.0,  TDuration::DurationType::V_64TH),       // 1/64
       TempoPattern("\uECAF",                     1.0/1920.0, TDuration::DurationType::V_128TH),      // 1/128
+      TempoPattern("\uECB1",                     1.0/1920.0, TDuration::DurationType::V_256TH),      // 1/256
+      TempoPattern("\uECB3",                     1.0/1920.0, TDuration::DurationType::V_512TH),      // 1/512
+      TempoPattern("\uECB5",                     1.0/1920.0, TDuration::DurationType::V_1024TH),     // 1/1024
```
Comment (Jojo-Schmitz) [inline libmscore/tempotext.cpp]: not sure these are needec

File: libmscore/tempotext.cpp [hunk 4]
```diff
@@ -165,6 +168,9 @@ static const TempoPattern tpSym[] = {
       TempoPattern("<sym>metNoteDoubleWhole</sym>",                                         1.0/7.5,   TDuration::DurationType::V_BREVE),    // double whole
       TempoPattern("<sym>metNote64thUp</sym>",                                              1.0/960.0, TDuration::DurationType::V_64TH),     // 1/64
       TempoPattern("<sym>metNote128thUp</sym>",                                             1.0/1920.0,TDuration::DurationType::V_128TH),    // 1/128
+      TempoPattern("<sym>metNote256thUp</sym>",                                             1.0/3840.0,TDuration::DurationType::V_256TH),    // 1/256
+      TempoPattern("<sym>metNote512thUp</sym>",                                             1.0/7680.0,TDuration::DurationType::V_512TH),    // 1/512
+      TempoPattern("<sym>metNote1024thUp</sym>",                                            1.0/15360.0,TDuration::DurationType::V_1024TH),  // 1/1024
```
Comment (Jojo-Schmitz) [inline libmscore/tempotext.cpp]: not sure these are needed

File: mscore/capella.cpp [hunk 5]
```diff
@@ -129,6 +129,10 @@ static void SetCapGraceDuration(Chord* chord,ChordObj* o)
             ((Chord*)chord)->setDurationType(TDuration::DurationType::V_128TH);
       else if (o->t == TIMESTEP::D256)
             ((Chord*)chord)->setDurationType(TDuration::DurationType::V_256TH);
+//      else if (o->t == TIMESTEP::D512)
+//            ((Chord*)chord)->setDurationType(TDuration::DurationType::V_512TH);
+//      else if (o->t == TIMESTEP::D1024)
+//            ((Chord*)chord)->setDurationType(TDuration::DurationType::V_1024TH);
```
Comment (Jojo-Schmitz) [inline mscore/capella.cpp]: Capella doesn't support such short durations, so no need for us to do.

File: mscore/capella.cpp [hunk 6]
```diff
@@ -2599,6 +2603,8 @@ Fraction BasicDurationalObj::ticks() const
             case TIMESTEP::D64:         len = Fraction(1,64);  break;
             case TIMESTEP::D128:        len = Fraction(1,128); break;
             case TIMESTEP::D256:        len = Fraction(1,256); break;
+//            case TIMESTEP::D512:        len = Fraction(1,512); break;
+//            case TIMESTEP::D1024:       len = Fraction(1,1024); break;
```
Comment (Jojo-Schmitz) [inline mscore/capella.cpp]: Capella doesn't support such short durations, so no need for us to do.

File: mscore/capella.h [hunk 7]
```diff
@@ -25,11 +25,11 @@
 
 namespace Ms {
 
-enum class TIMESTEP : char { D1, D2, D4, D8, D16, D32, D64, D128, D256, D_BREVE };
+enum class TIMESTEP : char { D1, D2, D4, D8, D16, D32, D64, D128, D256,/* D512, D1024,*/ D_BREVE };
```
Comment (Jojo-Schmitz) [inline mscore/capella.h]: Capella doesn't support such short durations, so no need for us to do.

File: mscore/capella.h [hunk 8]
```diff
@@ -25,11 +25,11 @@
...
 
 #if 0
 static const char* timeNames[] = { "1/1", "1/2", "1/4", "1/8", "1/16", "1/32", "1/64",
-      "1/128", "1/256", "breve" };
+      "1/128", "1/256", /*"1/512", "1/1024",*/ "breve" };
```
Comment (Jojo-Schmitz) [inline mscore/capella.h]: Capella doesn't support such short durations, so no need for us to do.

File: mscore/importmidi/importmidi_quant.cpp [hunk 9]
```diff
@@ -44,6 +44,15 @@ ReducedFraction quantValueToFraction(MidiOperations::QuantValue quantValue)
...
                   break;
+            case MidiOperations::QuantValue::Q_256:
+                  fraction = division / 64;
+                  break;
+            case MidiOperations::QuantValue::Q_512:
+                  fraction = division / 128;
+                  break;
+            case MidiOperations::QuantValue::Q_1024:
+                  fraction = division / 256;
+                  break;
```
Comment (Jojo-Schmitz) [inline mscore/importmidi/importmidi_quant.cpp]: Not sure we need such short quantization values

File: mscore/importmidi/importmidi_quant.cpp [hunk 10]
```diff
@@ -69,10 +78,17 @@ MidiOperations::QuantValue fractionToQuantValue(const ReducedFraction &fraction)
             quantValue = MidiOperations::QuantValue::Q_64;
       else if (fraction == division / 32)
             quantValue = MidiOperations::QuantValue::Q_128;
+      else if (fraction == division / 64)
+            quantValue = MidiOperations::QuantValue::Q_256;
+      else if (fraction == division / 128)
+            quantValue = MidiOperations::QuantValue::Q_512;
+      else if (fraction == division / 256)
+            quantValue = MidiOperations::QuantValue::Q_1024;
```
Comment (Jojo-Schmitz) [inline mscore/importmidi/importmidi_quant.cpp]: Not sure we need such short quantization values

File: mscore/prefsdialog.cpp [hunk 11]
```diff
@@ -511,14 +511,28 @@ void PreferenceDialog::updateValues(bool useDefaultValues)
...
+            shortestNoteIndex = 3;
+      else if (nn == MScore::division / 16)
+            shortestNoteIndex = 4;
+      else if (nn == MScore::division / 32)
+            shortestNoteIndex = 5;
+      else if (nn == MScore::division / 64) shortestNoteIndex = 6;
+      else if (nn == MScore::division / 128)
+            shortestNoteIndex = 7;
+      else if (nn == MScore::division / 256)
+            shortestNoteIndex = 8;
```
Comment (Jojo-Schmitz) [inline mscore/prefsdialog.cpp]: Not sure we need such short quantization values

File: mscore/prefsdialog.cpp [hunk 12]
```diff
@@ -1104,13 +1118,22 @@ void PreferenceDialog::apply()
...
+      switch (shortestNote->currentIndex()) {
+            case 0: ticks = MScore::division;       break;
+            case 1: ticks = MScore::division / 2;   break;
+            case 2: ticks = MScore::division / 4;   break;
+            case 3: ticks = MScore::division / 8;   break;
+            case 4: ticks = MScore::division / 16;  break;
+            case 5: ticks = MScore::division / 32;  break;
+            case 6: ticks = MScore::division / 64;  break;
+            case 7: ticks = MScore::division / 128; break;
+            case 8: ticks = MScore::division / 256; break;
```
Comment (Jojo-Schmitz) [inline mscore/prefsdialog.cpp]: Not sure we need such short quantization values

File: mscore/prefsdialog.ui [hunk 13]
```diff
@@ -3435,6 +3435,26 @@ Adjusting latency can help synchronize your MIDI hardware with MuseScore's inter
...
+            <item>
+             <property name="text">
+              <string>512th</string>
+             </property>
+            </item>
+            <item>
+             <property name="text">
+              <string>1024th</string>
+             </property>
+            </item>
```
Comment (Jojo-Schmitz) [inline mscore/prefsdialog.ui]: Not sure we need such short quantization values

File: mscore/importmidi/importmidi_operations.cpp [hunk 14]
```diff
@@ -74,6 +74,15 @@ static void setOperationsFromFile(const QString &fileName, Opers &opers)
...
                                           break;
+                                    case 6:
+                                          opers.quantValue.setDefaultValue(QuantValue::Q_256, false);
+                                          break;
+                                    case 7:
+                                          opers.quantValue.setDefaultValue(QuantValue::Q_512, false);
+                                          break;
+                                    case 8:
+                                          opers.quantValue.setDefaultValue(QuantValue::Q_1024, false);
+                                          break;
```
Comment (Jojo-Schmitz) [inline mscore/importmidi/importmidi_operations.cpp]: Not sure we need such short quantization values

File: mscore/importmidi/importmidi_operation.h [hunk 15]
```diff
@@ -19,7 +19,11 @@ enum class QuantValue {
       Q_16,
       Q_32,
       Q_64,
-      Q_128
+      Q_128,
+      Q_256,
+      Q_512,
+      Q_1024,
```
Comment (Jojo-Schmitz) [inline mscore/importmidi/importmidi_operation.h]: Not sure we need such short quantization values

File: mscore/importmidi/importmidi_operation.h [hunk 16]
```diff
@@ -19,7 +19,11 @@ enum class QuantValue {
...
+      Q_INVALID
```
Comment (Jojo-Schmitz) [inline mscore/importmidi/importmidi_operation.h]: this in turn should be usefull to have (and check for)

File: mscore/importmidi/importmidi_quant.cpp [hunk 17]
```diff
@@ -69,10 +78,17 @@ MidiOperations::QuantValue fractionToQuantValue(const ReducedFraction &fraction)
...
       else {
-            qDebug("unknown quant fraction %d/%d  division %d/%d", fraction.numerator(), fraction.denominator(),
-               division.numerator(), division.denominator());
-            Q_ASSERT_X(false, "Quantize::fractionToQuantValue", "Unknown quant fraction");
+            qDebug("Unknown quant fraction %d/%d in division %d/%d.",
+                   fraction.numerator(), fraction.denominator(),
+                   division.numerator(), division.denominator());
+            quantValue = MidiOperations::QuantValue::Q_INVALID;
```
Comment (Jojo-Schmitz) [inline mscore/importmidi/importmidi_quant.cpp]: this in turn should be usefull to have

File: mscore/importmidi/importmidi_quant.cpp [hunk 18]
```diff
@@ -81,7 +97,13 @@ MidiOperations::QuantValue fractionToQuantValue(const ReducedFraction &fraction)
 MidiOperations::QuantValue defaultQuantValueFromPreferences()
       {
       const auto fraction = ReducedFraction::fromTicks(preferences.getInt(PREF_IO_MIDI_SHORTESTNOTE));
-      return fractionToQuantValue(fraction);
+      MidiOperations::QuantValue quantValue = fractionToQuantValue(fraction);
+      if (quantValue == MidiOperations::QuantValue::Q_INVALID) {
+            qDebug("Unknown shortestNote value %d in preferences, defaulting to 16th.",
+                   preferences.getInt(PREF_IO_MIDI_SHORTESTNOTE));
+            quantValue = MidiOperations::QuantValue::Q_16;
```
Comment (Jojo-Schmitz) [inline mscore/importmidi/importmidi_quant.cpp]: this in turn should be usefull to have

File: mscore/importove.cpp [hunk 19]
```diff
@@ -989,6 +989,14 @@ TDuration OveNoteType_To_Duration(OVE::NoteType noteType){
...
                   break;
                   }
+//            case OVE::NoteType::Note_512: {
+//                  d.setType(TDuration::DurationType::V_512TH);
+//                  break;
+//                  }
+//            case OVE::NoteType::Note_1024: {
+//                  d.setType(TDuration::DurationType::V_1024TH);
+//                  break;
+//                  }
```
Comment (Jojo-Schmitz) [inline mscore/importove.cpp]: Not sure whether Overtures supposes those

File: mscore/ove.cpp [hunk 20]
```diff
@@ -1776,6 +1776,8 @@ void NoteContainer::setNoteType(NoteType type) {
             case NoteType::Note_64:
             case NoteType::Note_128:
             case NoteType::Note_256: {
+//            case NoteType::Note_512:
+//            case NoteType::Note_1024: {
```
Comment (Jojo-Schmitz) [inline mscore/ove.cpp]: Not sure whether Overtures supposes those

File: mscore/ove.cpp [hunk 21]
```diff
@@ -1935,6 +1937,14 @@ int NoteContainer::getDuration() const {
...
                   break;
                   }
+//            case NoteType::Note_512: {
+//                  duration = (int) NoteDuration::D_512;
+//                  break;
+//                  }
+//            case NoteType::Note_1024: {
+//                  duration = (int) NoteDuration::D_1024;
+//                  break;
+//                  }
```
Comment (Jojo-Schmitz) [inline mscore/ove.cpp]: Not sure whether Overtures supposes those

File: mscore/ove.h [hunk 22]
```diff
@@ -351,6 +351,8 @@ enum class NoteType : char {
       Note_64          = 0x7,
       Note_128         = 0x8,
       Note_256         = 0x9,
+//      Note_512         = 0xa,
+//      Note_1024        = 0xb,
```
Comment (Jojo-Schmitz) [inline mscore/ove.h]: Not sure whether Overtures supposes those

File: mscore/prefsdialog.cpp [hunk 23]
```diff
@@ -511,14 +511,28 @@ void PreferenceDialog::updateValues(bool useDefaultValues)
...
+      else if (nn == MScore::division / 32)
+            shortestNoteIndex = 5;
+      else if (nn == MScore::division / 64) shortestNoteIndex = 6;
+      else if (nn == MScore::division / 128)
+            shortestNoteIndex = 7;
+      else if (nn == MScore::division / 256)
+            shortestNoteIndex = 8;
+      else {
+            qDebug("Unknown shortestNote value of %d, defaulting to 16th", nn);
+            shortestNoteIndex = 2;
```
Comment (Jojo-Schmitz) [inline mscore/prefsdialog.cpp]: this though should be usefull

File: mscore/prefsdialog.cpp [hunk 24]
```diff
@@ -1104,13 +1118,22 @@ void PreferenceDialog::apply()
...
+            case 3: ticks = MScore::division / 8;   break;
+            case 4: ticks = MScore::division / 16;  break;
+            case 5: ticks = MScore::division / 32;  break;
+            case 6: ticks = MScore::division / 64;  break;
+            case 7: ticks = MScore::division / 128; break;
+            case 8: ticks = MScore::division / 256; break;
+            default: {
+                  qDebug("Unknown index for shortestNote: %d, defaulting to 16th",
+                         shortestNote->currentIndex());
+                  ticks = MScore::division / 4;
```
Comment (Jojo-Schmitz) [inline mscore/prefsdialog.cpp]: this though should be useful

File: libmscore/shadownote.cpp [hunk 25]
```diff
@@ -86,6 +86,15 @@ SymId ShadowNote::getNoteFlag() const
...
                   break;
+            case TDuration::DurationType::V_256TH:
+                  flag = computeUp() ? SymId::flag256thUp : SymId::flag256thDown;
+                  break;
+            case TDuration::DurationType::V_512TH:
+                  flag = computeUp() ? SymId::flag512thUp : SymId::flag512thDown;
+                  break;
+            case TDuration::DurationType::V_1024TH:
+                  flag = computeUp() ? SymId::flag1024thUp : SymId::flag1024thDown;
+                  break;
```
Comment (Jojo-Schmitz) [inline libmscore/shadownote.cpp]: Those shadow notes (seen during note input) have way too long stems. No idea how to fix that.^
Those stems seem to depend on the Musical font used, with Bravura even longer durations than those added here have quite long stems, with Emmentaler just these added ones.

File: mscore/prefsdialog.ui (hunk 13 above)
Comment (Marr11317) [reply mscore/prefsdialog.ui:3504]: yes please, it is always useful when the tempo is very slow.

File: libmscore/shadownote.cpp [hunk 26]
```diff
@@ -151,9 +160,10 @@ void ShadowNote::draw(QPainter* painter) const
             QPointF pos;
             pos.rx() = up == 1 ? (noteheadWidth - (lw / 2)) : lw / 2;
             qreal yOffset = up == 1 ? symStemUpSE(_notehead).y() * magS() : symStemDownNW(_notehead).y() * magS();
-            if(flag != SymId::lastSym) {
-                  pos.ry() -= up * (symHeight(flag) + (posDot.y() != 0 ? posDot.y() + spatium() : 0) + 0.5*spatium());
-                  painter->drawLine(QLineF(pos.x(), yOffset, pos.x(), pos.y() - up * (yOffset + lw/2)));
+            if (flag != SymId::lastSym) {
+                  qreal flagHeight = (_duration.type() < TDuration::DurationType::V_256TH) ? symHeight(flag) : symHeight(flag) / 2;
```
Comment (Jojo-Schmitz) [inline libmscore/shadownote.cpp:164]: Those shadow notes (seen during note input) had way too long stems, for those added durations.
Those stems seem to depend on the Musical font used, with Bravura's even longer durations than those added here have quite long stems, with Emmentaler just these added ones.
With this change it looks better, but this seems to be quick a hack?

**Comment by Jojo-Schmitz** [comment]
Fixed the new icons, they do now fit into their buttons (but are showing smaller than the others).
Sort of fixed the overly long stem for the 'shadow' notes (those shown during note entry) too.

File: libmscore/tempotext.cpp [hunk 27]
```diff
@@ -120,6 +120,9 @@ static const TempoPattern tp[] = {
       TempoPattern("\uECA0",                     1.0/7.5,    TDuration::DurationType::V_BREVE),      // double whole
       TempoPattern("\uECAD",                     1.0/960.0,  TDuration::DurationType::V_64TH),       // 1/64
       TempoPattern("\uECAF",                     1.0/1920.0, TDuration::DurationType::V_128TH),      // 1/128
+      TempoPattern("\uECB1",                     1.0/1920.0, TDuration::DurationType::V_256TH),      // 1/256
```
Comment (anatoly-os) [inline libmscore/tempotext.cpp:123]: `1.0 / 3840.0` then?

Comment (Jojo-Schmitz) [reply libmscore/tempotext.cpp:123]: oops

**Comment by anatoly-os** [comment]
Oops...

**Comment by anatoly-os** [comment]
@Jojo-Schmitz please create a separate PR to address the review. Sorry...

**Comment by Jojo-Schmitz** [comment]
done, see  #5974

**Pull request merged**